packages/vm/src/sections.rs

```diff
@@ -0,0 +1,157 @@
+use crate::conversion::to_u32;
+use crate::errors::VmResult;
+
+/// Decodes sections of data into multiple slices.
+///
+/// Each encoded section is suffixed by a section length, encoded as big endian uint32.
+///
+/// See also: `encode_section`.
+#[allow(dead_code)]
+pub fn decode_sections(data: &[u8]) -> Vec<&[u8]> {
+    let mut result: Vec<&[u8]> = vec![];
+    let lengths = extract_lengths(data);
+    let mut start = 0;
+    for length in lengths.iter().rev() {
+        result.push(&data[start..start + length]);
+        start += length + 4;
+    }
+    result
+}
```

Comment:
LGTM. You could also use the `split_tail` implementation and read the tail until the rest is empty. Then you need to reverse the result, which is cheap for a vector of references.

Reply:
Ah, nevermind. This is a zero-copy decoder in contrast to `split_tail`, which is built to return onwned result (shinks first section and copies all other sections). So you should keep the independent implementation.

You could optionally use the same strategy: decode and push to result in one loop instead of creating the positions first and then using them. However, this is just a matter of taste.